### PR TITLE
Fournir `meeting_point_id` dans le job de synchro ANTS

### PR DESCRIPTION
**Pull request by francois-ferrandis**
Suite de #4940

# Contexte

L'ANTS nous informe que le `meeting_point_id` doit être passé afin qu'ils puissent avoir un suivi des bonnes pratiques des mairies : 

> C'est à partir du meetingpoint_id que nous identifions les mairies/communes qui font appel aux routes. Pour le cas de la route get status, cela nous permet de savoir quelles mairies vérifient le statut du numéro de pré-demande lors de la prise de rdv.
> Avec un traçage de tous les appels, nous pourrons ainsi tracer le cycle complet de la prise de rendez-vous.
> Par ailleurs, en cas de dysfonctionnement cela nous permet également de plus facilement identifier le meetingpoint_id qui pose problème.

cf https://gitlab.com/france-titres/rendez-vous-mairie/interaction-avec-les-editeurs/-/issues/101

Nous envoyons déjà ce paramètre dans quasiment tous nos appels, sauf les appels à `status` où nous envoyons ce paramètre, mais vide. Nous pensions que c’était une erreur de la prat de l’ANTS avant qu’ils ne nous répondent.

# Solution

Dans cette première PR pour nous aligner, on ajoute le passage du paramètre `meeting_point_id` à l’appel status dans le job de synchro ANTS.

Ce job de synchronisation prend en paramètre un numéro de pré-demande. 
On a maintenant besoin d’un `meeting_point_id` dès le début de ce job pour appeler `status`.

On va chercher ce `meeting_point_id` à différents endroits et on prend le premier : 
- il peut être passé explicitement en paramètre du job (utile lors de suppression de participations)
- on cherche un RDV à venir dont un user a ce numéro de prédemande et on utilise son `lieu_id` le cas échéant.
- on cherche un RDV passé dont un user a ce numéro de prédemande et on utilise son `lieu_id` le cas échéant
- on raise une erreur si on n’en n’a pas trouvé

### CODE REVIEW

File: app/lib/ants_api.rb [hunk 1]
```diff
@@ -20,16 +20,13 @@ class ApiRequestError < StandardError; end
...
-        meeting_point_id: nil, # required but unused cf https://github.com/betagouv/rdv-service-public/pull/4940
-      }
+    def status(ants_pre_demande_number:, meeting_point_id: nil, timeout: nil)
+      params = { application_ids: ants_pre_demande_number, meeting_point_id: }
       response_body = request(:get, "status", params:, timeout:)
       response_body.fetch(ants_pre_demande_number)
     end
 
-    def create(ants_pre_demande_number:, meeting_point:, management_url:, appointment_date:, meeting_point_id: nil)
+    def create(ants_pre_demande_number:, meeting_point:, management_url:, appointment_date:, meeting_point_id:)
```
Comment (adipasquale) [inline app/lib/ants_api.rb:29]: note: ce passage du paramètre `meeting_point_id:` en obligatoire pour `create` et `delete` déborde un tout petit peu du cadre de cette PR. 
On envoie déjà ce paramètre lorsqu’on appelle ces méthodes. 
Dorénavant, on ne pourra plus faire autrement, c’est un ajout de filet de sécurité.

File: spec/jobs/ants/sync_appointment_job_spec.rb [hunk 2]
```diff
@@ -65,13 +65,24 @@
...
-    it "ne créé pas d’appointment" do
-      allow(AntsApi).to receive(:status)
-        .with(hash_including(ants_pre_demande_number: "A123456789"))
-        .and_return({ "status" => "validated", "appointments" => [] })
+    it "ne fait rien et raise" do
+      expect(AntsApi).not_to receive(:status)
       expect(AntsApi).not_to receive(:delete)
       expect(AntsApi).not_to receive(:create)
-      described_class.perform_now(ants_pre_demande_number: "A123456789")
+      expect { described_class.new.perform(ants_pre_demande_number: "A123456789") }.to raise_error Ants::MissingMeetingPointId
```
Comment (adipasquale) [inline spec/jobs/ants/sync_appointment_job_spec.rb:72]: Dans ce contexte « Synchro pour un ants_pre_demande_number correspondant à un usager ayant un RDV mais qui n’est pas un RDV ANTS », le comportement change : 

- avant le job n’échouait pas et faisait une requête à `status`
- maintenant le job échoue et ne fait pas de requête à `status`

J’ai relu le `Ants::AppointmentSerializerAndListener` et ce cas ne devrait pas arriver.

Comment (adipasquale) [reply spec/jobs/ants/sync_appointment_job_spec.rb:72]: au passage, le `perform_now` semble rescue l’exception, tandis que le `new.perform` la laisse bien remonter pour cette spec

File: app/models/concerns/ants/appointment_serializer_and_listener.rb [hunk 3]
```diff
@@ -14,7 +14,15 @@ module AppointmentSerializerAndListener
...
-        Ants::AppointmentSerializerAndListener.mark_for_sync([participation.rdv], obsolete_ants_pre_demande_number: participation.user.ants_pre_demande_number)
+        if participation.user.ants_pre_demande_number.present?
+          Ants::AppointmentSerializerAndListener.mark_for_sync(
+            [participation.rdv],
+            obsolete_ants_data: {
+              pre_demande_number: participation.user.ants_pre_demande_number,
+              meeting_point_id: participation.rdv.lieu_id,
+            }
+          )
+        end
```
Comment (adipasquale) [inline app/models/concerns/ants/appointment_serializer_and_listener.rb:25]: ces changements permettent de gérer le cas d’une suppression de participation : 
lorsque le job de synchro tourne avec le numéro ANTS de l’usager retiré, la participation a déjà été supprimée. Il est donc impossible de retrouver le RDV ANTS correspondant dans le job de synchro. Ça bloque donc pour l’appel à `status` au début de ce job, qui a maintenant besoin d’un lieu_id, celui du RDV.

J’ai ici un peu élargi le mécanisme de stockage transitoire de données sur le RDV via un attr_accessor `:obsolete_ants_data`.

File: app/models/concerns/ants/appointment_serializer_and_listener.rb [hunk 4]
```diff
@@ -14,7 +14,15 @@ module AppointmentSerializerAndListener
         Ants::AppointmentSerializerAndListener.mark_for_sync(user.rdvs) if user.saved_change_to_ants_pre_demande_number?
       end
       Participation.before_commit do |participation|
-        Ants::AppointmentSerializerAndListener.mark_for_sync([participation.rdv], obsolete_ants_pre_demande_number: participation.user.ants_pre_demande_number)
+        if participation.user.ants_pre_demande_number.present?
```
Comment (adipasquale) [inline app/models/concerns/ants/appointment_serializer_and_listener.rb:17]: ce `if present?` était avant fait plus bas dans `enqueue_sync_for_marked_record` via `rdv.obsolete_ants_pre_demande_number.present?`

File: spec/support/ants_stubs.rb [hunk 5]
```diff
@@ -1,6 +1,8 @@
-def stub_request_ants_status(application_id)
+def stub_request_ants_status(application_id, meeting_point_id: nil)
```
Comment (adipasquale) [inline spec/support/ants_stubs.rb:1]: on rendra ce paramètre obligatoire dans une PR successive

File: app/jobs/ants/sync_appointment_job.rb [hunk 6]
```diff
@@ -1,20 +1,25 @@
...
+      @ants_pre_demande_number = ants_pre_demande_number
 
-      ants_status = AntsApi.status(ants_pre_demande_number: stripped_ants_pre_demande_number, timeout: 4)
+      ants_status = AntsApi.status(
+        ants_pre_demande_number: stripped_ants_pre_demande_number,
+        meeting_point_id: meeting_point_id || find_meeting_point_id,
+        timeout: 4
+      )
 
       return false unless ants_status["status"] == "validated"
```
Comment (francois-ferrandis) [inline app/jobs/ants/sync_appointment_job.rb:24]: Ça n'a pas été introduit dans cette PR mais je remarque qu'on a `return false` ici et `return true` plus bas, alors que la valeur de retour n'est jamais utilisée nulle part.

File: app/jobs/ants/sync_appointment_job.rb [hunk 7]
```diff
@@ -24,18 +29,8 @@ def perform(ants_pre_demande_number:)
...
-        .order(id: :desc) # choix arbitraire pour éviter un comportement aléatoire
-        .first
-
-      # on ne fait rien si les infos sont déjà identiques
-      return true if ants_appointments == [rdv&.serialize_for_ants_api]
+      # on ne fait rien si les infos sont déjà identiques (chacune des deux listes peuvent être vides ici)
+      return true if ants_appointments == [upcoming_rdv_serialized_to_ants_appointment]
 
       # on déclenche la suppression des appointments existants dans tous les cas, qu’il s’agisse d’une mise à jour ou d’une suppression
       # en effet l’API de l’ANTS ne permet pas de faire de mises à jour, on fait donc un delete puis un update
```
Comment (francois-ferrandis) [inline app/jobs/ants/sync_appointment_job.rb]: Au passage, ce commentaire est incorrect, non ?

```suggestion
      # en effet l’API de l’ANTS ne permet pas de faire de mises à jour, on fait donc un delete puis un create
```

File: app/jobs/ants/sync_appointment_job.rb [hunk 8]
```diff
@@ -24,18 +29,8 @@ def perform(ants_pre_demande_number:)
...
-        .where(users: { ants_pre_demande_number: [stripped_ants_pre_demande_number, ants_pre_demande_number] })
-        .where.not(status: Rdv::CANCELLED_STATUSES)
-        .where("starts_at >= ?", Time.zone.now)
-        .order(id: :desc) # choix arbitraire pour éviter un comportement aléatoire
-        .first
-
-      # on ne fait rien si les infos sont déjà identiques
-      return true if ants_appointments == [rdv&.serialize_for_ants_api]
+      # on ne fait rien si les infos sont déjà identiques (chacune des deux listes peuvent être vides ici)
+      return true if ants_appointments == [upcoming_rdv_serialized_to_ants_appointment]
```
Comment (francois-ferrandis) [inline app/jobs/ants/sync_appointment_job.rb]: Je me demande dans quel cas on a deux listes vides. :thinking:  Je vais réfléchir à haute voix. On est donc dans un cas où :

- le numéro de pré-demande est valide mais nous n'avons pas (encore) déclaré de RDV pour ce numéro
- nous n'avons pas d'upcoming RDV (`upcoming_rdv` est `nil`)
- nous avons quand-même trouvé un RDV ANTS pour ce numéro (méthode de fallback dans `find_meeting_point_id`)

J'ai du mal à imaginer un cas où ça arrive, et toi ? :smiley_cat: 

En tout cas le code est raisonnable : si on a pas de RDV à venir, on a rien à déclarer, et donc le job peut s'arrêter là ! :heavy_check_mark:

Comment (francois-ferrandis) [reply app/jobs/ants/sync_appointment_job.rb]: Remarque purement technique : il faudrait appeler `#compact` sur l'array pour éviter une valeur `[nil]`.

Je ne sais pas si ça vaut le coup d'ajouter un test vu que c'est un edge case... et en le disant, ça donne envie d'ajouter ce test ! 

```ruby
context "Synchro pour un RDV ANTS dans le passé" do
# ...

    context "quand l'ANTS ne nous retourne aucun RDV" do
      it "n'appelle ni delete ni create" do
        allow(AntsApi).to receive(:status)
          .with(hash_including(ants_pre_demande_number: "A123456789", meeting_point_id: lieu.id.to_s))
          .and_return(
            {
              "status" => "validated",
              "appointments" => [],
            }
          )
        expect(AntsApi).not_to receive(:delete)
        expect(AntsApi).not_to receive(:create)
        described_class.perform_now(ants_pre_demande_number: "A123456789")
      end
    end
```

En lançant le test, on voit qu'il passe malgré la comparaison erronée, car en fait `ants_appointments.each` ne fait rien et on s'arrête ligne 45 à `return unless upcoming_rdv`.

File: app/models/concerns/ants/appointment_serializer_and_listener.rb [hunk 9]
```diff
@@ -14,7 +14,15 @@ module AppointmentSerializerAndListener
...
+            [participation.rdv],
+            obsolete_ants_data: {
+              pre_demande_number: participation.user.ants_pre_demande_number,
+              meeting_point_id: participation.rdv.lieu_id,
+            }
+          )
+        end
       end
       Lieu.before_commit do |lieu|
         Ants::AppointmentSerializerAndListener.mark_for_sync(lieu.rdvs) if lieu.saved_change_to_name?
```
Comment (francois-ferrandis) [inline app/models/concerns/ants/appointment_serializer_and_listener.rb:28]: Cette ligne me terrifie, mais c'est un sujet pour un autre jour. :dash:

File: app/models/concerns/ants/appointment_serializer_and_listener.rb (hunk 3 above)
Comment (francois-ferrandis) [reply app/models/concerns/ants/appointment_serializer_and_listener.rb:25]: Super solution, c'est simple et malin !

**Review by francois-ferrandis — COMMENTED** [review]
Je valide (même si je ne peux pas le faire formellement car c'est ma propre PR)

**Review by adipasquale — APPROVED** [review]
